src/physics/em/detail/RayleighInteractor.i.hh

```diff
@@ -15,54 +18,123 @@ namespace detail
+                                       const ParticleTrackView&      particle,
+                                       const Real3&                  direction,
+                                       ElementId                     el_id)
+
     : shared_(shared)
     , inc_energy_(particle.energy().value())
-    , inc_direction_(inc_direction)
-    , allocate_(allocate)
+    , inc_direction_(direction)
+    , element_id_(el_id)
 {
     CELER_EXPECT(inc_energy_ >= this->min_incident_energy()
                  && inc_energy_ <= this->max_incident_energy());
-    CELER_EXPECT(particle.particle_id() == shared_.gamma_id); // XXX
-    CELER_NOT_IMPLEMENTED("Rayleigh scattering");
+    CELER_EXPECT(particle.particle_id() == shared_.gamma_id);
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Sample using the XXX model.
+ * Sample the Rayleigh scattering angle using the G4LivermoreRayleighModel
+ * and G4RayleighAngularGenerator of Geant4 6.10.
  */
 template<class Engine>
 CELER_FUNCTION Interaction RayleighInteractor::operator()(Engine& rng)
 {
-    // Allocate space for XXX (electron, multiple particles, ...)
-    Secondary* secondaries = this->allocate_(0); // XXX
-    if (secondaries == nullptr)
-    {
-        // Failed to allocate space for a secondary
-        return Interaction::from_failure();
-    }
-
-    // XXX sample
-    (void)sizeof(rng);
+    real_type energy = inc_energy_.value();
 
     // Construct interaction for change to primary (incident) particle
     Interaction result;
-    result.action      = Action::scattered;                     // XXX
-    result.energy      = units::MevEnergy{inc_energy_.value()}; // XXX
-    result.direction   = inc_direction_;
-    result.secondaries = {secondaries, 1}; // XXX
+    result.action = Action::scattered;
+    result.energy = units::MevEnergy{inc_energy_.value()};
+
+    // Sample direction for a given Z: G4RayleighAngularGenerator
+    ItemIdT item_id = ItemIdT{element_id_.get()};
+
+    Real3 pn = shared_.params.data_n[item_id];
+    Real3 pb = shared_.params.data_b[item_id];
+    Real3 px = shared_.params.data_x[item_id];
+
+    SampleInput input = this->evaluate_weight_and_prob(energy, pb, pn, px);
+
+    real_type                          cost;
+    UniformRealDistribution<real_type> u01(0, 1.0);
+
+    do
+    {
+        unsigned  index = 0;
+        real_type x     = u01(rng);
+
+        if (x > input.prob[0])
+        {
+            index = (x <= input.prob[0] + input.prob[1]) ? 1 : 2;
+        }
+
+        real_type w = input.weight[index];
+        real_type n = pn[index];
+        real_type b = pb[index];
+
+        n = 1.0 / n;
+
+        // Sampling of scattering angle
+        real_type y = w * u01(rng);
+        if (y < num_limit())
+        {
+            x = y * n * (1. + 0.5 * (n + 1.) * y * (1. - (n + 2.) * y / 3.));
+        }
+        else
+        {
+            x = std::exp(-n * std::log(1. - y)) - 1.0;
```

Comment:
Done and droped decimals in elsewhere - did not known that using 1.0 involves an unnecessary double conversion (actually I thought the opposite way) !

Reply:
Yeah, it's really tricky. For a real_type of `v`:
- `v + 1 / 3` will be `v + 1` because of order-of operations (really bad!)
- `v + 0.3` will promote `v` to double-precision and do the addition in double-precision, then demote `v` back to float (bad!)
- `v + 0.5` shockingly on some compilers (GCC10 host) seems to be optimized so that it does the addition directly as a floating-point value. Same for `+ 1.0` -- I guess because 1 and .5 are exactly the same in float/double and can be converted back and forth and guaranteed to be correct.
- `v + 1` promotes the integer to a double so is fine
- `v + real_type(0.3)` will truncate 0.3 to 0.3f at compile time as needed and do the addition in single-precision.

So it looks like you're actually right in the particular case of 1.0 🤔 The more you know...

Reply:
And even worse `std::log(1. - f)` promotes `f` to double (then does double-precision log) and doesn't use the 1.0 shortcut! So super confusing. Easier to stick with `int` when the order-of-op doesn't get you in trouble, or add `real_type` around the literals when it might 😄

Reply:
This is more complicated than I though.  Do we need something like one<T>(), half<T>() and so on for commonly used hard coded numbers?